src/wp-includes/html-api/class-wp-html-tag-processor.php

```diff
@@ -2732,19 +3188,25 @@ public function has_self_closing_flag(): bool {
 	 * @return bool Whether the current tag is a tag closer.
 	 */
 	public function is_tag_closer(): bool {
-		return (
-			self::STATE_MATCHED_TAG === $this->parser_state &&
-			$this->is_closing_tag &&
+		if ( self::STATE_MATCHED_TAG !== $this->parser_state ) {
+			return false;
+		}
 
-			/*
-			 * The BR tag can only exist as an opening tag. If something like `</br>`
-			 * appears then the HTML parser will treat it as an opening tag with no
-			 * attributes. The BR tag is unique in this way.
-			 *
-			 * @see https://html.spec.whatwg.org/#parsing-main-inbody
-			 */
-			'BR' !== $this->get_tag()
-		);
+		switch ( $this->parsing_namespace ) {
+			case 'html':
+				/*
+				 * The BR tag can only exist as an opening tag. If something like `</br>`
+				 * appears then the HTML parser will treat it as an opening tag with no
+				 * attributes. The BR tag is unique in this way.
+				 *
+				 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+				 */
+				return $this->is_closing_tag && 'BR' !== $this->get_tag();
+
+			case 'math':
+			case 'svg':
+				return $this->is_closing_tag;
```

Comment:
@sirreal why remove `has_self_closing_flag()` from this? if it's present, I was considering that a closing tag.

maybe the terminology is skewed when it comes to self-closing elements. it _does_ close the tag, but it is simultaneously an opener. to me, `is_tag_closer()` captures that, but maybe that's confusing.

perhaps this is best handled in `expects_closer()`

Reply:
It's unclear exactly how these should behave. This change fixed an issue where this condition was not being entered on self-closing tags:

https://github.com/WordPress/wordpress-develop/blob/f66ba57cd96c6d0e38e81d90bf91134423d7844e/src/wp-includes/html-api/class-wp-html-processor.php#L4018-L4042

If this condition is included, it becomes difficult to know if a tag is intended to create an element, so the condition above would have to become something like `! $this->is_tag_closer() && ! $this->has_self_closing_flag()` in order to know that it _is_ a tag opener.

---

I'm not convinced one way or another.

On one hand, it's a tag opener and a tag closer, so reporting that it's a tag closer is correct.
On the other hand, it feels like they're _essentially_ openers and not closers. They happen to be self-close, but that's not what we're really asking.

Maybe we need a new method, or maybe that one place I mentioned above just needs to be updated to account for self-closers being reported as closers.

Reply:
after talking about this I think you're right. this should, inevitably, be handled in `expects_closer()`, not in `is_tag_closer()`